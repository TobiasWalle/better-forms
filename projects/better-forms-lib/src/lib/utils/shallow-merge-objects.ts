export function shallowMergeObjects<T>(objects: T[]): T {
  return objects.reduce((object, nextObject) => {
    return {
      ...object,
      ...nextObject
    };
  }, {});
}
