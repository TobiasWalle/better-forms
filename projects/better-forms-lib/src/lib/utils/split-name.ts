export function splitName(path: string): string[] {
  return path.split('.');
}
