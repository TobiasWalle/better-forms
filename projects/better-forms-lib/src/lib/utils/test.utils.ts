import { ControlValueAccessor } from '@angular/forms';

export function createMockValueAccessor(): ControlValueAccessor {
  return {
    writeValue: jest.fn(),
    registerOnChange: jest.fn(),
    registerOnTouched: jest.fn(),
    setDisabledState: jest.fn(),
  };
}
