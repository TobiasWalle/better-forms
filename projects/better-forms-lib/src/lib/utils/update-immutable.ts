const immutable: any = require('object-path-immutable');

export function setImmutable<T>(original: T, path: string[], newValue: any): T {
  return immutable.set(original, path, newValue);
}

export function getByPath<T>(object: T, path: string[]): any {
  return path.reduce((nextObject: any, key: string) => {
    if (key in nextObject) {
      return nextObject[key];
    }
    throw new Error(`Couldn't find "${path}" on object. "${key}" doesn not exists on ${nextObject}.`);
  }, object);
}

