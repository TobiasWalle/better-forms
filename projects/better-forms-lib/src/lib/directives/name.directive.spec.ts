import { ControlValueAccessor } from '@angular/forms';
import { createMockValueAccessor } from '../utils/test.utils';
import { FormDirective } from './form.directive';
import { NameDirective } from './name.directive';

describe('NameDirective', () => {
  let directive!: NameDirective;

  const mockForm: FormDirective = {
    register: jest.fn(),
    unregister: jest.fn(),
  } as any;
  const mockValueAccessors: ControlValueAccessor[] = [
    createMockValueAccessor()
  ];

  beforeEach(() => {
    directive = new NameDirective(
      mockForm,
      mockValueAccessors,
      null
    );
  });

  it('should create an instance', () => {
    expect(directive).toBeTruthy();
  });

  it('should throw an error if the value accessors are null or empty', () => {
    expect(() => new NameDirective(mockForm, null, null)).toThrowErrorMatchingSnapshot();
    expect(() => new NameDirective(mockForm, [], null)).toThrowErrorMatchingSnapshot();
  });

  describe('#ngOnInit', () => {
    it('should register with name', async () => {
      directive.betterFormName = 'my-name';
      await directive.ngOnInit();

      expect(mockForm.register).toHaveBeenCalledWith('my-name', mockValueAccessors[0], []);
    });

    it('should throw an error if name is not set', async () => {
      directive.betterFormName = undefined;

      await expect(directive.ngOnInit()).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('#ngOnInit', () => {
    it('should unregister', () => {
      directive.betterFormName = 'my-name';
      directive.ngOnDestroy();

      expect(mockForm.unregister).toHaveBeenCalledWith('my-name');
    });

    it('should throw an error if name is not set', () => {
      directive.betterFormName = undefined;

      expect(() => directive.ngOnDestroy()).toThrowErrorMatchingSnapshot();
    });
  });
});
