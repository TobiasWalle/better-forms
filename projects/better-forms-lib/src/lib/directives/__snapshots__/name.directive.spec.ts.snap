// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameDirective #ngOnInit should throw an error if name is not set 1`] = `"The \\"betterFormName\\" Input is required"`;

exports[`NameDirective #ngOnInit should throw an error if name is not set 2`] = `"The \\"betterFormName\\" Input is required"`;

exports[`NameDirective should throw an error if the value accessors are null or empty 1`] = `"No ValueAccessor found"`;

exports[`NameDirective should throw an error if the value accessors are null or empty 2`] = `"No ValueAccessor found"`;
